### DISCUSSION TITLE
`null` as input and default value, but not output

### DISCUSSION THREAD

**Discussion by lukasvice** [open-circle/valibot#376] [Q&A]
I'd like to have a schema with the following behavior (e.g. for a number input field):

Input type: `number | null`
Output type: `number`
Default: `null`

So I tried the following:
- `v.nullable(v.number())` -> Output type is `number | null`; Default is `undefined`
- `v.nullable(v.number(), 0)` -> Default is `0`
- I also tried `v.transform` but I didn't find a way to do it.

Thanks for your advice!

**Comment by fabian-hiller** [comment]
Where does the `number` output come from if someone passes `null` as input and the default should also be `null`? If the default is `null`, I see no way to turn that into a number for the output.

**Reply by lukasvice** [reply]
Right, the idea was that if `null` is passed, it will result in an error.

For example, imagine a number input field that returns a number or `null` if the input is not a number. If you want to start with an empty input, you can set `null` as the default value.
But in the end you should be required to enter a number.

**Reply by fabian-hiller** [reply]
I think what you are trying to do is technically impossible with a schema library. On the one hand you want the schema to pass `null` as a default value, and on the other hand you want to throw an error for `null` values. This contradicts itself. But maybe I'm wrong. Have you ever done the same thing with another schema library?

However, if you separate your form logic from the schema and only use the schema for the final validation, it is possible. In this case, you can use `v.number()` for the field in your schema.

**Reply by lukasvice** [reply]
> I think what you are trying to do is technically impossible with a schema library. On the one hand you want the schema to pass null as a default value, and on the other hand you want to throw an error for null values. This contradicts itself. But maybe I'm wrong.

I see your point. Perhaps the default value in this case is more of a fallback value? For example, how could you define a schema with a fallback value of `0` (if the input is `null`) but with a default value (provided by `getDefaults`) of `50`? I could imagine something like `z.default(v.nullable(z.number(), 0), 50)`? But that's off topic and maybe worth another discussion :)

> However, if you separate your form logic from the schema and only use the schema for the final validation, it is possible. In this case, you can use v.number() for the field in your schema.

Yes, that would be possible, but I thought there might be a better way, so I could also manage the default values with the schema library.

> Have you ever done the same thing with another schema library?

Sort of. See the discussion in this [Zod issue](https://github.com/colinhacks/zod/issues/1206).
My last [two comments](https://github.com/colinhacks/zod/issues/1206#issuecomment-1597126130) show my approach. But there are two problems with it: It's very verbose and Zod can't return default values like `getDefaults`. I wanted to see if Valibot would be able to help me with this or if it's really better to separate the validation from the form logic.

**Reply by fabian-hiller** [reply]
> I see your point. Perhaps the default value in this case is more of a fallback value?

That could be possible with `z.fallback(v.nullable(z.number(), 0), 50)`.

> Yes, that would be possible, but I thought there might be a better way

Does the following code work for you? To allow a default value of `null` to be retrieved via `getDefaults`, we could change the code of `nullable`.

```ts
const Schema = v.transform(
  v.nullable(v.number()),
  (input) => input || 0,
  v.number()
);
```

**Reply by fabian-hiller** [reply]
I am also interested in what your dream API would be. Maybe we can make it happen.

**Reply by dboune** [reply]
@fabian-hiller I have the same or similar set of issues/needs. I've only recently begun using schema-oriented validation libraries, and of course, I'm smitten with the ability to infer an input and output type. I've been using TypeBox and Valibot in different situations to great effect.

In the places I use Valibot, I regularly need to accept one or more of [null, an empty string, or undefined] as input but prohibit output until a stricter set of criteria is met, meaning my output should always meet specification, or I shouldn't get an output but an error instead. Using value validations isn't always the most desirable solution, whereas schema validations would meet the need perfectly.

I most need a way to allow null, undefined, or literal('') (or some combination thereof) as an input, but given a validation, know that an error output and thus a non-compliant output (if we were using `safeParse`) are inevitable if the validation does not pass. `literal('')` is a pretty special case, but it can be important for form validation with controlled inputs, even for non-string values.

In my mind, something like the following should be all I need...

```ts
type RepeatTypeInput = Input<typeof RepeatType>;
type RepeatType = Output<typeof RepeatType>;
const RepeatType = nullable(
  picklist(['daily', 'weekly', 'monthly'], 'Please select a repeat type.'),
);
```

With a type result of...

```ts
type RepeatTypeInput = "daily" | "weekly" | "monthly" | null
type RepeatType = "daily" | "weekly" | "monthly"
```

Unfortunately, I can only do this for this case with a simple string with `nullable` and a default value of `''`. Then, I have to create a custom validation pipe for the list of valid values where needed, which means my input type can't be as specific as a downstream type requires. In other situations, like arrays of strings, I can use an empty array as a default to nullable and have it work out as expected.


If I wanted `null` to be valid for input and output, I'd do this instead.

```ts
const RepeatType = union([
  null_(),
  picklist(['daily', 'weekly', 'monthly'], 'Please select a valid repeat type.'),
], 'Invalid repeat type.');
```

Presently, nullable is either a form of shorthand for the above or a way to coerce null into a value that passes schema validation.

`nullable` doesn't have to be the solution, but it would be nice to have a solution that only affects the input type, thus permitting "acceptable invalid values" where needed.

**Reply by dboune** [reply]
This is the workaround. Of course, it gets a lot more verbose when you have larger structures, variations, etc. Additionally, keeping separate default sets. It would be advantageous to work all of this into one definition.

```ts
export type RepeatTypeInput = Input<typeof RepeatType> | null;
export type RepeatType = Output<typeof RepeatType>;
export const RepeatType = picklist(['daily', 'weekly', 'monthly'], 'Please select a repeat type.'),
```

And in the more complex cases, I have to just keep an entirely separate typing for input.

**Reply by fabian-hiller** [reply]
Can't you achieve your goal with the following?

```ts
import * as v from 'valibot';

const RepeatSchema = v.nullable(
  v.picklist(['daily', 'weekly', 'monthly']),
  'daily'
);

type RepeatInput = v.Input<typeof RepeatSchema>; // "daily" | "weekly" | "monthly" | null
type RepeatOutput = v.Output<typeof RepeatSchema>; // "daily" | "weekly" | "monthly"
```

**Reply by dboune** [reply]
@fabian-hiller That depends on whether or not we want null to be treated as a valid value rather than only a valid input.

Please see this sandbox for two examples that may help clarify: https://codesandbox.io/p/devbox/react-mui-react-hook-form-valibot-hddfsq

In the second example (`App2.tsx`), I'm using a custom utility that makes null a permissible input without allowing it to pass validation. I could maintain a separate input type, but this solution is convenient. I realize it's a rather useless bit of kit outside of modifying the input type, but it's difficult to do this otherwise for very complex forms.

I've created multiple variations in my project, including one that permits a literal empty string input, and they have worked wonderfully for me.

It is understandable if this is out of scope for Valibot, though if it is easy enough to accommodate, it could be helpful for those working with form validation.

**Reply by fabian-hiller** [reply]
Thanks for the code example! I now understand your use case better.  As I understand it, you only need `null` for the `defaultValues` of `useForm`. Why do you need `null` for this at all? Why can't it be undefined? Personally, I would try to remove `null` from the schema and instead of `RepeatFormSchemaInput` I would write `DeepPartial<RepeatFormSchemaInput>` since that is the type `useForm` expects.

**Reply by dboune** [reply]
The use of `literal('')`, or `null` in special cases like MUI Joy's Select component, rather than `undefined` stems from the referenced warning presented by React if `undefined` or `null` are used as the initial value for a field that later becomes controlled.

https://github.com/facebook/react/blob/06e410ec608820b65d81536bb4630228b4e04d48/packages/react-dom-bindings/src/client/ReactDOMComponent.js#L1487-L1492

In Joy's Select component, unlike in a plain React component, `null` is treated as "no option selected" and is the value output by onChange when no value is selected. IIRC literal('') is acceptable as an initial value, but the component may modify it to be `null`.

As for React's behavior concerning controlled vs uncontrolled components, it's a decision that was made early on and is unlikely ever to change.

Using `DeepPartial` for the type fed to `useForm` could work. However, from a DX perspective, that may not be optimal as `undefined`, in particular, is broadly problematic for the controlled component use case. When working with the form's values during the operation of the form, the types fed to `useForm` become important. For the controlled form use case, it seems best to guard against `undefined` during form input and operation.

**Reply by fabian-hiller** [reply]
I understand your situation, but it seems to me that the problem is more related to your form and/or component library. For a schema library, it logically makes no sense to have `null` as an input type when it is not a valid input type, because it produces an issue. Nevertheless, I am happy to get such feedback. Do you think there is anything Valibot should change or improve to make it better?

**Reply by fabian-hiller** [reply]
You can also create your own `DeepNullable` type to transform the input type of Valibot to optionally allow `null` for any input. Reach out if I should send you sample code.

**Reply by dboune** [reply]
Thanks, @fabian-hiller! I appreciate your taking the time to read these discussion threads, seriously consider the needs of your users, and provide your input. I'm not alone in feeling very positive about your approach.

A `DeepNullable` utility could also work when the entire form comprises only `Select` components or other components accepting `null` as "empty". I would need to be targeted. Not every field should allow `null` input, even as an **acceptable invalid input**.

The `nullableInput`, `emptyInput`, `optionalInput`, and `nullishInput` I've added for myself work well enough to create a clean, simple, unified experience. It's already saved me many headaches, and it's very nice that this can be done easily.

I can only agree that this need is specific to the concept of _form validation_,  with additional specificity for React and component libraries like MUI (Joy is an official variant of MUI) and the usage pattern of react-hook-form, which are currently among the most commonly used. Your user base will likely be composed of a fair number of people using this combination and some portion of those confronting similar situations. As developers, we are all looking for ways to simplify our work while maintaining quality.

For the sake of conversation, Valibot presents itself as a schema library offering the ability to "Validate unknown data." Indeed, the `parse` method takes `unknown` as input. Valibot is a schema library, but I also viewed it as a validation library, including schema and value validation, so I generally expect the potential for Input to be less strict than the definition and expected output. As I said before, I adore the type inference capabilities tools like Valibot provide and what that affords me as a developer in a TypeScript environment.

My queries seem to center around my misuse (or unexpected creative use 😁), but from the perspective of form validation (recognizing that another situation may have a different set of needs), I took the `Input` type as the type to be used for expected inputs to my form, which for a form must include **acceptable invalid input** as a core concept due to the nature of a form, even where we don't want to assign a default, fallback, or transform. Output requirements are naturally more strict, even considering **schema validation**.

On the `Infer types` page, Valibot states...

> You are probably interested in input type only in special cases. I recommend you to use the output type in most cases.

At least in the form validation process, **this is that special case for me!** I can imagine other instances where we want to assert schema via typing first (which I also need to do here, but with more flexibility) and then primarily validate values. However, if I am in a position where I cannot fully trust the schema type, then utilizing `Input` is rather dangerous.

For that polar opposite, if I were using Valibot to validate an HTTP POST body, I'd be uninterested in `Input` as I couldn't trust any assertion it makes. Handling that kind of input as `unknown` is the only appropriate way to view it until the schema can be validated.

Thanks for creating Valibot! I was using TypeBox everywhere previously, and while it is a stellar tool, I'm not using its JSON schema capabilities other than value validation, and it's quite a bit more work to process and customize the error messages it provides when considering form validation in particular. Valibot offers a sensible API with excellent OOTB capabilities, making it an easy choice. It might interest you that the TypeBox Workbench convinced me that Valibot was the right choice, as my interfaces that compiled cleanly to TypeBox code also compiled cleanly to Valibot code but not to any other supported tool.

**Reply by fabian-hiller** [reply]
Thank you for your detailed answer and your kind words. This kind of feedback is very important for further development of Valibot. I found two more solutions that might work for you.

In the first one you keep your schema as it is (without `nullableInput`). Also, you do not provide any default values to the form. This way the value of your field will be `undefined`. When passing the value to your MUI component, write the following `value ?? null` to prevent passing an undefined value. If you try this, let me know if it works.

The other solution requires a new function called `input` that I could add to Valibot. `input` takes a schema as first argument and allows to transform it's input type via it's generics. I have not investigated the implementation, but this is what the API could look like.

```ts
import * as v from 'valibot';

const RepeatSchema = v.input<null>(v.picklist(['daily', 'weekly', 'monthly']));

type RepeatInput = v.Input<typeof RepeatSchema>; // "daily" | "weekly" | "monthly" | null
type RepeatOutput = v.Output<typeof RepeatSchema>; // "daily" | "weekly" | "monthly"
```

**Reply by dboune** [reply]
Defining the schema without any notion of `optional`, `nullable`, or `nullish`, the answer is a partial yes. It's a valid suggestion.

From a defensive programming perspective, this could be a good idea. However, one has to be very careful to handle all possible cases.. we may not be able to rely on `||` because a numeric zero could be a valid value, but `??` may not either be perfect as we may need to cover empty string for some cases, so our tests might become complex and prone to requiring frequent customization thereby making it difficult to componentize frequently used inputs. I've tried to avoid doing too much with `value` and `onChange` this way as it tends to be error-prone or requires too much situation-dependent variation.

As a form becomes more complex, such as dealing with variants, create and edit modes, etc., we start to face more challenges. It may indeed become necessary to utilize `Partial` too broadly, which can lead to confusion later on.

Additionally, if using Eslint with `typescript-eslint/strict-type-checked` we will encounter errors testing for values not included in the type, which I realize is a choice. I prefer to be very explicit end to end.

So, the `input` function option with the flexible type generic seems perfect for my needs. It allows me to be very explicit about what constitutes an **acceptable invalid input**, which greatly simplifies ensuring form inputs meet component requirements, whatever those might happen to be. This also helps a great deal when dealing with database output where `null` is a common value, and we need to make sure that values are transformed appropriately.

**Reply by fabian-hiller** [reply]
Thanks for your feedback. Feel free to create an issue for an `input` and maybe an `output` function. I am not sure if I want to add such functionality, but the issue will give us a dedicated space to discuss and get opinions from other developers.

**Reply by TeChn4K** [reply]
I was looking for this exact discussion all over existing validation/schema libraries! Thanks I'm not alone anymore!

My needs are very similar : 
- initialize form inputs with null values (fields are empty at start)
- some fields are optionals, but some others required
- have an "all nullable inputs" schema (for react-hook-form defaultValues & watch)
- have an "strict output" schema (for onSubmit)

I'll take this discussion further for this use case. Thanks all.
